src/Spectre.Service.Tests/Loaders/DatasetLoaderTest.cs

```diff
@@ -0,0 +1,109 @@
+        private readonly string _remoteDir = "remote";
+
+        private readonly string _fileDir = $"{TestContext.CurrentContext.TestDirectory} + "
+                                           + @"\..\..\..\..\..\test_files";
+        [OneTimeSetUp]
+        public void SetUp()
+        {
+            DependencyResolver.AddModule(new MockModule());
+
+            var localDirFull = Path.Combine(_rootDir, _localDir);
+            var remoteDirFull = Path.Combine(_rootDir, _remoteDir);
+            var correctDataset = File.ReadAllText(Path.Combine(_fileDir, "small-test.txt"));
+
+            _mockFileSystem = DependencyResolver.GetService(typeof(IFileSystem)) as MockFileSystem;
+
+            _mockFileSystem.AddFile(Path.Combine(localDirFull, "local_correct.txt"), new MockFileData(correctDataset));
+            _mockFileSystem.AddFile(Path.Combine(remoteDirFull, "remote_correct.txt"), new MockFileData(correctDataset));
+            _mockFileSystem.AddFile(Path.Combine(localDirFull, "local_incorrect.txt"), new MockFileData(textContents: "incorrect_data"));
+            _mockFileSystem.AddFile(Path.Combine(remoteDirFull, "remote_incorrect.txt"), new MockFileData(textContents: "incorrect_data"));
+
+            _rootConfig = new DataRootConfig(localDirFull, remoteDirFull);
+            _datasetLoader = new DatasetLoader(_rootConfig);
+        }
+
+        [Test]
+        public void ReturnsFromCorrectNameLocal()
+        {
+            Assert.IsNotNull(anObject: _datasetLoader.GetFromName(name: "local_correct.txt"),
+                message: "Loader did not manage to load local file.");
+        }
+
+        [Test]
+        public void ReturnsFromCorrectNameRemote()
+        {
+            Assert.IsNotNull(anObject: _datasetLoader.GetFromName(name: "remote_correct.txt"),
+                message: "Loader did not manage to load remote file.");
+        }
+
+        [Test]
+        public void ThrowsOnIncorrectName()
+        {
+            Assert.Throws<DatasetNotFoundException>(code: () => _datasetLoader.GetFromName(name: "invalid_name"),
+                message: "Loader accepted invalid file name.");
+        }
+
+        [Test]
+        public void ThrowsOnIncorrectFileContents()
+        {
+            Assert.Throws<DatasetFormatException>(code: () => _datasetLoader.GetFromName(name: "local_incorrect.txt"),
+                message: "Loader did not manage to load remote file.");
+        }
+
+        [Test]
+        public void DeletesIncorrectFilesFromLocal()
+        {
+            try
+            {
+                _datasetLoader.GetFromName(name: "remote_incorrect.txt");
+            }
+            catch (Exception)
+            {
+                // ignored
+            }
+            var result = _mockFileSystem.AllFiles.FirstOrDefault(predicate: file => file.Contains(value: Path.Combine(_localDir, "remote_incorrect.txt")));
+            Assert.IsNull(result, message: "Loader leaves copies of incorrect files in local directory.");
```

Comment:
Does remote remain untouched? It is good, but I just want to confirm.

Reply:
Yeah, the file persists on the remote side.